### PR TITLE
Dradis WPScan

### PR DESCRIPTION
**Pull request by ethicalhack3r**
### Summary

Creates a WPScan upload plugin.

### Other Information

1. I'd like to create evidence for some of the issues but not sure what to use as the `node` value.

2. Is it possible to set severities when imported?

### Copyright assignment

Collaboration is difficult with commercial closed source but we want
to keep as much of the OSS ethos as possible available to users
who want to fix it themselves.

In order to unambiguously own and sell Dradis Framework commercial
products, we must have the copyright associated with the entire
codebase. Any code you create which is merged must be owned by us.
That's not us trying to be a jerks, that's just the way it works.

Please review the [CONTRIBUTING.md](https://github.com/dradis/dradis-ce/blob/master/CONTRIBUTING.md)
file for the details.

You can delete this section, but the following sentence needs to
remain in the PR's description:

> I assign all rights, including copyright, to any future Dradis
> work by myself to Security Roots.

### CODE REVIEW

File: lib/dradis/plugins/wpscan/importer.rb [hunk 1]
```diff
@@ -12,16 +12,140 @@ def import(params={})
...
+      vulnerabilities.each do |vulnerability|
+        logger.info { "Adding vulnerability: #{vulnerability['title']}" }
+
+        vulnerability_template = template_service.process_template(template: 'vulnerability', data: vulnerability)
+        content_service.create_issue(text: vulnerability_template, id: vulnerability['wpvulndb_id'])
+      end
+
+      # Add bespoke/config vulnerabilities to Dradis
+      #
+      # TODO: Would be better to add the URL & passwords as evidence.
```
Comment (dormi) [inline lib/dradis/plugins/wpscan/importer.rb]: An idea that may improve on this:

at the beginning of the import, you may create a node with the sacaned target_url as label (similar to what it is going on here: https://github.com/dradis/dradis-acunetix/blob/master/lib/dradis/plugins/acunetix/importer.rb#L36)

Then you can create the issues as you are doing, but you can also create evidence  to show that those issues are happening in that node (target_url).

This way of later we import another wpscan file with another target_url, existing issues won't be duplicated, but the evidence will be created for the new target_url node.

Comment (ethicalhack3r) [reply lib/dradis/plugins/wpscan/importer.rb]: Thanks for your feedback!

I'll implement the changes tonight/tomorrow sometime.

File: templates/vulnerability.sample [hunk 2]
```diff
@@ -0,0 +1,481 @@
+{
```
Comment (dormi) [inline templates/vulnerability.sample]: No need to add the whole json here? just the relevant part to build the vulnerability may be enough?

Something like:
```
{ 
   "title":"qTranslate 2.5.34 - Setting Manipulation CSRF",
    "fixed_in":null,
    "references":{ 
       "cve":[ 
          "2013-3251"
        ],
       "wpvulndb":[ 
          "6846"
        ]
     }
  }
```

Comment (ethicalhack3r) [reply templates/vulnerability.sample]: I could trim it down, but I think the sample file also helps in testing, and the more real world it is the better for testing purposes. But I really don't mind either way.

File: lib/dradis/plugins/wpscan/importer.rb (hunk 1 above)
Comment (dormi) [reply lib/dradis/plugins/wpscan/importer.rb]: Note that this suggestion only makes sense if we can find the evidence that relates each issue (vulnerability) to the node (target_url).

**Comment by ethicalhack3r** [comment]
Implemented your feedback, thanks! 

Couple of questions:

1. Is there a way to add an `issue` to a `node` without `evidence`?

2. Is there a way to add severity during import?

**Comment by dormi** [comment]
> Is there a way to add an issue to a node without evidence?
No, I don't think this is how Dradis is supposed to work. There are issues, there are nodes, and evidence says with nodes are affect by each issues. 
Can you give an example of your use case?

**Comment by ethicalhack3r** [comment]
> Can you give an example of your use case?

To link the vulnerability to the node (`target_url`) even if there's no suitable "evidence".

**Comment by dormi** [comment]
> Is there a way to add severity during import?

Yes. For example, to add it to an issue:
 - add `vulnerability.severity` to `templates/vulnerability.fields`
- add 
```
#[Severity]#
%vulnerability.severity%
``` 
to `templates/vulnerability.template`

Then, how to populate `vulnerability.severity` ?
I don't know the response to that.
But if it is a value that cannot be just picked up from the json and needs some extra logic, maybe it's worth mapping the 
vulnerability to its own class, as we done in other plugins: https://github.com/dradis/dradis-netsparker/blob/master/lib/netsparker/vulnerability.rb
That new class would implement a `severity` method.
If we follow that path, then we must also fine tune `field_processor.rb` to use that new `Vulnerability` class: https://github.com/dradis/dradis-netsparker/blob/master/lib/dradis/plugins/netsparker/field_processor.rb

I think that knowing from where the severity value is coming will help to improve this answer.

**Comment by ethicalhack3r** [comment]
Thanks! Can `vulnerability.severity` be any value I choose, or does it have to be a string, integer or float?

For example

- `high`, `medium`, `low`
- `0`, `5`
- `4.3`, `7`, `10`

**Comment by ethicalhack3r** [comment]
OK, I think for now the changes are good. I can always add severity, etc. at a later date.

Let me know if you have any feedback on the latest changes, please :)

**Comment by ethicalhack3r** [comment]
This is ready to merge if you are happy with the current state. Thanks!

**Comment by dormi** [comment]
Hi @ethicalhack3r.
This looks good. As you said we can always stuff at a later date.

Would it be possible to add some minimal specs?
Something like: https://github.com/dradis/dradis-nmap/blob/master/spec/nmap_upload_spec.rb

A part from that this initial version looks ready to merge. 
Thanks a lot.

**Comment by ethicalhack3r** [comment]
I get this error when trying to run `rspec`, do you know where I need to place my plugin code so that it includes the Dradis classes/methods?

```
wpscan upload plugin
  Importer
    raises an error note when the json is not valid (FAILED - 1)

Failures:

  1) wpscan upload plugin Importer raises an error note when the json is not valid
     Failure/Error: @content_service = Dradis::Plugins::ContentService.new(plugin: plugin)
     
     NoMethodError:
       undefined method `new' for Dradis::Plugins::ContentService:Module
     # ./spec/fixtures/files/wpscan_upload_spec.rb:15:in `block (3 levels) in <top (required)>'

Finished in 0.009 seconds (files took 1.15 seconds to load)
1 example, 1 failure
```

**Comment by ethicalhack3r** [comment]
Hi! Just thought I'd follow up on this, thanks!

**Comment by dormi** [comment]
Sorry @ethicalhack3r, I think I pointed you to a broken spec example. 
Fixing.
This one works for me: https://github.com/dradis/dradis-nikto/blob/master/spec/nikto_upload_spec.rb
Note that it uses:
`Dradis::Plugins::ContentService::Base`

**Comment by dormi** [comment]
@ethicalhack3r the specs I mentioned are being fixed in https://github.com/dradis/dradis-nmap/pull/17

If you push your current implemented specs I may be able to help.

**Comment by ethicalhack3r** [comment]
Thanks!

I'm getting this failure after updating the specs:

```
Failures:

  1) wpscan upload plugin Importer raises an error note when the json is not valid
     Failure/Error: data = MultiJson.decode(file_content)
     
     NameError:
       uninitialized constant Dradis::Plugins::Wpscan::Importer::MultiJson
     # ./lib/dradis/plugins/wpscan/importer.rb:12:in `import'
     # ./spec/fixtures/files/wpscan_upload_spec.rb:24:in `block (3 levels) in <top (required)>'
```

Do you know what the issue might be?

**Comment by dormi** [comment]
I think we need https://github.com/ethicalhack3r/dradis-wpscan/pull/1

File: spec/fixtures/files/wpscan_upload_spec.rb [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        content_service: @content_service,
+      )
+    end
+
+    it "raises an error note when the json is not valid" do
+      # Run the import
+      p @importer.import(file: 'spec/fixtures/files/invalid.json')
+    end
+
+    # it "creates nodes, issues, notes and an evidences as needed" do
```
Comment (dormi) [inline spec/fixtures/files/wpscan_upload_spec.rb]: @ethicalhack3r this should be uncommented, yes?

Comment (ethicalhack3r) [reply spec/fixtures/files/wpscan_upload_spec.rb]: Yea, was just debugging at the time and didn't want to lose my local changes.

I had a quick look earlier but came across another couple of issues. I will have another look tomorrow, and take a look at some other plugin spec examples, and hopefully get it finished.

**Comment by ethicalhack3r** [comment]
Hey! Sorry for the delay on this.

I'm struggling to get the specs working, and can't find any working spec examples that could help me.

**Comment by dormi** [comment]
I was thinking on doing something similar to the specs I fixed for `dradis-nmap`: https://github.com/dradis/dradis-nmap/blob/fix-specs/spec/nmap_upload_spec.rb

**Comment by ethicalhack3r** [comment]
I couldn't get the `dradis-nmap` specs to run.

```
$ rspec
The git source https://github.com/dradis/dradis-plugins.git is not yet checked out. Please run `bundle install` before trying to start your application
```

```
$ bundle install
Fetching https://github.com/dradis/dradis-plugins.git
Fetching gem metadata from https://rubygems.org/..............
Fetching gem metadata from https://rubygems.org/.
Resolving dependencies...
Bundler could not find compatible versions for gem "bundler":
  In Gemfile:
    bundler (~> 1.6)

  Current Bundler version:
    bundler (2.0.2)
This Gemfile requires a different version of Bundler.
Perhaps you need to update Bundler by running `gem install bundler`?

Could not find gem 'bundler (~> 1.6)' in any of the relevant sources:
  the local ruby installation
```

```
$ bundle _1.6_ install
[...]
Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

    current directory: /Users/ryandewhurst/.rvm/gems/ruby-2.6.3/gems/nokogiri-1.10.7/ext/nokogiri
/Users/ryandewhurst/.rvm/rubies/ruby-2.6.3/bin/ruby -I /Users/ryandewhurst/.rvm/rubies/ruby-2.6.3/lib/ruby/site_ruby/2.6.0 -r ./siteconf20191216-44495-1i1us8v.rb extconf.rb ''
extconf failedno implicit conversion of nil into String

Gem files will remain installed in /Users/ryandewhurst/.rvm/gems/ruby-2.6.3/gems/nokogiri-1.10.7 for inspection.
Results logged to /Users/ryandewhurst/.rvm/gems/ruby-2.6.3/extensions/x86_64-darwin-18/2.6.0/nokogiri-1.10.7/gem_make.out
An error occurred while installing nokogiri (1.10.7), and Bundler cannot continue.
Make sure that `gem install nokogiri -v '1.10.7'` succeeds before bundling.
```

```
$ gem install nokogiri -v '1.10.7'
[...]
NoMethodError: undefined method `spec' for nil:NilClass
An error occurred while installing dradis-plugins (3.15.0), and Bundler cannot continue.
```

**Comment by dormi** [comment]
`dradis-nmap` spec were fixed in https://github.com/dradis/dradis-nmap/pull/17

Another plugin with working specs: https://github.com/dradis/dradis-nexpose

`bundler >= 1.6` can be removed: https://github.com/dradis/dradis-nexpose/commit/8d0b381dbffe70396e659590d6191bf63055c4d3

**Comment by dormi** [comment]
@ethicalhack3r trying to add specs myself I think that we may need to better document "adding specs" to how to build an import plugin docs.

Meanwhile I've managed to get some basic specs working in https://github.com/ethicalhack3r/dradis-wpscan/pull/2

**Comment by ethicalhack3r** [comment]
Hey @dormi,

I hope you had a great Christmas and new Year.

Can this be merged, or do you require more specs?

Thanks,
Ryan

**Comment by dormi** [comment]
ping @brianp @aapomm 
I think this can be merged now.

**Pull request merged**